Report: pnpm audit broken (410) — blocks pre-commit hook

## Summary

Since **2026-04-15**, `pnpm audit` fails with HTTP 410 because the npm registry retired both legacy audit endpoints. This breaks our pre-commit hook (`lefthook.yml` → `pre-commit → audit`) and blocks every commit on this repo.

## Upstream tracking

- **pnpm/pnpm#11265** — https://github.com/pnpm/pnpm/issues/11265 (OPEN, opened 2026-04-15)
- Affects pnpm **10.x** (incl. our pinned `10.33.0`) **and** **11.0.0-rc.0** — upgrading the `packageManager` does not fix it.
- Behavior is intermittent (some users report it briefly working again, then failing), which matches what we observe in CI/local.

## Reproduction (this repo)

```bash
$ pnpm audit --audit-level=high
 ERR_PNPM_AUDIT_BAD_RESPONSE  The audit endpoint (at
 https://registry.npmjs.org/-/npm/v1/security/audits/quick) responded with 410:
 {"error":"This endpoint is being retired. Use the bulk advisory endpoint instead.
 See the following docs for more info: https://api-docs.npmjs.com/#tag/Audit"}.
 Fallback endpoint (...) responded with 410: ...
```

Same error surfaces at the end of `pnpm latest` (`latest:update` script) and inside `pnpm git -y` via the lefthook `audit` step — blocking commits that have already passed prettier, cspell, eslint, and type-check.

## Impact on our workflow

- `fullrun` flow stops after lint/test pass, at the commit step.
- `pnpm latest` exits non-zero even when `update --latest` succeeds (package state is fine; only the trailing audit call fails).
- CLAUDE.md rules forbid `--no-verify` without explicit user approval, so commits must be unblocked another way per incident.

## Options (deferred — needs decision per incident)

- (a) Commit with `--no-verify` when the hook fails (explicit, case-by-case).
- (b) Remove / make-advisory the `audit` command in [lefthook.yml](lefthook.yml) pre-commit.
- (c) Replace with a non-registry audit (e.g. `pnpm dlx better-npm-audit` / `osv-scanner` / `pnpm dlx audit-ci` against a local advisory source).
- (d) Wait for upstream fix (pnpm/pnpm#11265) and do nothing.

## Action

Watch pnpm/pnpm#11265. Once a pnpm release ships with the bulk advisory endpoint, bump `packageManager` in `package.json` and close this issue.

## Out of scope for this issue

- Removing `pnpm.overrides` entries (CLAUDE.md forbids without user approval).
- Changing `packageManager` pin — 11.0.0-rc.0 is also affected per upstream comments.

---

Filed at user request; no code changes in this issue.

--- a/.github/workflows/ci.yml
+++ b/.github/workflows/ci.yml
@@ -48,8 +48,16 @@ jobs:
       - name: Install dependencies
         run: pnpm install --frozen-lockfile --ignore-scripts
 
+      - name: Setup Go for osv-scanner
+        uses: actions/setup-go@v5
+        with:
+          go-version: stable
+
+      - name: Install osv-scanner
+        run: go install github.com/google/osv-scanner/v2/cmd/osv-scanner@latest
+
       - name: Run security audit
-        run: pnpm audit --audit-level low
+        run: osv-scanner --lockfile=pnpm-lock.yaml
 
       - name: Run spell check
         run: pnpm exec cspell . --dot

--- a/.github/workflows/ci.yml
+++ b/.github/workflows/ci.yml
@@ -54,6 +54,8 @@ jobs:
           go-version: stable
 
       - name: Install osv-scanner
+        # Intentionally unpinned: a vulnerability scanner should always pull the
+        # newest detector + advisory set. The determinism tradeoff is accepted.
         run: go install github.com/google/osv-scanner/v2/cmd/osv-scanner@latest
 
       - name: Run security audit

--- a/cspell.config.yaml
+++ b/cspell.config.yaml
@@ -84,6 +84,8 @@ words:
   - noreferrer
   - nosniff
   - octocat
+  - osv
+  - ghcr
   - opencollective
   - paraglide
   - pexels

--- a/lefthook.yml
+++ b/lefthook.yml
@@ -6,7 +6,7 @@ pre-commit:
       run: pnpm run prevent-main-commit
 
     audit:
-      run: pnpm audit
+      run: pnpm audit:security
 
     cspell:
       glob: '*.{ts,tsx,js,jsx,svelte,md,mdx,html,css,scss,yml,yaml}'

--- a/package.json
+++ b/package.json
@@ -1,7 +1,7 @@
 {
 	"name": "joshuafolkken-com",
 	"private": true,
-	"version": "0.156.0",
+	"version": "0.157.0",
 	"type": "module",
 	"packageManager": "pnpm@10.33.0+sha512.10568bb4a6afb58c9eb3630da90cc9516417abebd3fabbe6739f0ae795728da1491e9db5a544c76ad8eb7570f5c4bb3d6c637b2cb41bfdcdb47fa823c8649319",
 	"engines": {
@@ -73,11 +73,12 @@
 		"version:minor": "tsx scripts/bump-version.ts minor",
 		"version:major": "tsx scripts/bump-version.ts major",
 		"version:patch": "tsx scripts/bump-version.ts patch",
-		"latest": "pnpm latest:corepack && pnpm latest:update && pnpm audit",
+		"latest": "pnpm latest:corepack && pnpm latest:update && pnpm audit:security",
 		"latest:corepack": "corepack use pnpm@latest",
 		"latest:update": "pnpm update --latest",
 		"prep": "tsx scripts/prep.ts",
 		"overrides:check": "tsx scripts/overrides-check.ts",
+		"audit:security": "tsx scripts/security-audit.ts",
 		"check:upstream": "tsx scripts/check-upstream-drift.ts",
 		"issue:prep": "tsx scripts/issue-prep.ts",
 		"git": "tsx scripts/git-workflow.ts",

--- a/scripts/security-audit-logic.ts
+++ b/scripts/security-audit-logic.ts
@@ -1,0 +1,29 @@
+const BINARY_NAME = 'osv-scanner'
+const LOCKFILE_PATH = 'pnpm-lock.yaml'
+
+const INSTALL_INSTRUCTIONS = [
+	'',
+	'Install options:',
+	'  macOS:  brew install osv-scanner',
+	'  Go:     go install github.com/google/osv-scanner/v2/cmd/osv-scanner@latest',
+	'  Docker: docker run --rm -v "$PWD:/src" ghcr.io/google/osv-scanner --lockfile=/src/pnpm-lock.yaml',
+	'',
+	'Docs: https://google.github.io/osv-scanner/',
+].join('\n')
+
+function build_scanner_arguments(lockfile_path: string): ReadonlyArray<string> {
+	return [`--lockfile=${lockfile_path}`]
+}
+
+function format_missing_binary_error(binary_name: string): string {
+	return `${binary_name} is not installed.\n${INSTALL_INSTRUCTIONS}`
+}
+
+const security_audit_logic = {
+	BINARY_NAME,
+	LOCKFILE_PATH,
+	build_scanner_arguments,
+	format_missing_binary_error,
+}
+
+export { security_audit_logic }

--- a/scripts/security-audit.test.ts
+++ b/scripts/security-audit.test.ts
@@ -1,0 +1,108 @@
+import { readFileSync } from 'node:fs'
+import path from 'node:path'
+import { describe, expect, it } from 'vitest'
+import { security_audit_logic } from './security-audit-logic'
+
+const OSV_SCANNER = 'osv-scanner'
+const PNPM_LOCKFILE = 'pnpm-lock.yaml'
+const RETIRED_AUDIT = 'pnpm audit'
+const RETIRED_AUDIT_RUN = `run: ${RETIRED_AUDIT}`
+const RETIRED_AUDIT_FRAGMENT = `${RETIRED_AUDIT} `
+
+function load_file(relative_path: string): string {
+	return readFileSync(path.resolve(process.cwd(), relative_path), 'utf8')
+}
+
+describe('security_audit_logic constants', () => {
+	it('targets the osv-scanner binary', () => {
+		expect(security_audit_logic.BINARY_NAME).toBe(OSV_SCANNER)
+	})
+
+	it('targets the pnpm lockfile at repo root', () => {
+		expect(security_audit_logic.LOCKFILE_PATH).toBe(PNPM_LOCKFILE)
+	})
+})
+
+describe('security_audit_logic.build_scanner_arguments', () => {
+	it('returns --lockfile with the given path', () => {
+		expect(security_audit_logic.build_scanner_arguments(PNPM_LOCKFILE)).toEqual([
+			`--lockfile=${PNPM_LOCKFILE}`,
+		])
+	})
+
+	it('preserves custom lockfile paths unchanged', () => {
+		expect(security_audit_logic.build_scanner_arguments('app/pnpm-lock.yaml')).toEqual([
+			'--lockfile=app/pnpm-lock.yaml',
+		])
+	})
+})
+
+describe('security_audit_logic.format_missing_binary_error', () => {
+	const message = security_audit_logic.format_missing_binary_error(OSV_SCANNER)
+
+	it('names the missing binary', () => {
+		expect(message).toContain(`${OSV_SCANNER} is not installed`)
+	})
+
+	it('includes the brew install instruction', () => {
+		expect(message).toContain('brew install osv-scanner')
+	})
+
+	it('includes the go install instruction', () => {
+		expect(message).toContain('go install github.com/google/osv-scanner')
+	})
+
+	it('includes the Docker fallback', () => {
+		expect(message).toContain('docker run')
+	})
+
+	it('links to the osv-scanner docs', () => {
+		expect(message).toContain('https://google.github.io/osv-scanner/')
+	})
+})
+
+describe('package.json audit wiring', () => {
+	const manifest = JSON.parse(load_file('package.json')) as { scripts?: Record<string, string> }
+	// eslint-disable-next-line dot-notation -- noPropertyAccessFromIndexSignature forbids dot access on Record values
+	const latest_script = manifest.scripts?.['latest'] ?? ''
+
+	it('registers audit:security as a tsx wrapper entry', () => {
+		expect(manifest.scripts?.['audit:security']).toBe('tsx scripts/security-audit.ts')
+	})
+
+	it('uses audit:security inside the latest script', () => {
+		expect(latest_script).toContain('pnpm audit:security')
+	})
+
+	it('no longer leaves the retired pnpm audit at the tail of the latest script', () => {
+		expect(latest_script.trim().endsWith(RETIRED_AUDIT)).toBe(false)
+		expect(latest_script).not.toContain(RETIRED_AUDIT_FRAGMENT)
+	})
+})
+
+describe('lefthook.yml pre-commit audit wiring', () => {
+	const content = load_file('lefthook.yml')
+	const lines = content.split('\n').map((line) => line.trim())
+
+	it('runs audit:security on pre-commit', () => {
+		expect(content).toContain('run: pnpm audit:security')
+	})
+
+	it('does not call the retired pnpm audit command from the pre-commit hook', () => {
+		expect(lines).not.toContain(RETIRED_AUDIT_RUN)
+	})
+})
+
+describe('.github/workflows/ci.yml audit wiring', () => {
+	const content = load_file('.github/workflows/ci.yml')
+	const lines = content.split('\n').map((line) => line.trim())
+
+	it('runs osv-scanner against the pnpm lockfile', () => {
+		expect(content).toContain(`osv-scanner --lockfile=${PNPM_LOCKFILE}`)
+	})
+
+	it('does not call the retired pnpm audit command from CI', () => {
+		expect(lines).not.toContain('run: pnpm audit --audit-level low')
+		expect(lines).not.toContain(RETIRED_AUDIT_RUN)
+	})
+})

--- a/scripts/security-audit.ts
+++ b/scripts/security-audit.ts
@@ -1,0 +1,29 @@
+#!/usr/bin/env node
+import { spawnSync } from 'node:child_process'
+import { security_audit_logic } from './security-audit-logic'
+
+const FAILURE_EXIT_CODE = 1
+
+function is_binary_available(binary_name: string): boolean {
+	const result = spawnSync(binary_name, ['--version'], { stdio: 'ignore' })
+
+	return result.error === undefined
+}
+
+function main(): never {
+	const { BINARY_NAME, LOCKFILE_PATH, build_scanner_arguments, format_missing_binary_error } =
+		security_audit_logic
+
+	if (!is_binary_available(BINARY_NAME)) {
+		console.error(format_missing_binary_error(BINARY_NAME))
+		process.exit(FAILURE_EXIT_CODE)
+	}
+
+	const result = spawnSync(BINARY_NAME, [...build_scanner_arguments(LOCKFILE_PATH)], {
+		stdio: 'inherit',
+	})
+
+	process.exit(result.status ?? FAILURE_EXIT_CODE)
+}
+
+main()

